### ISSUE TITLE
setMaxTimeInSeconds does not work properly

### ISSUE THREAD

**Issue by pfolch** [google/or-tools#1196]
Hi all,
First of all thanks for the great work behind this library. I found a bug related with the search time limits on the SatSolver, with the next chunk of code (in Java) I will try to show it.

```java
public static final long numIter = 20000;
public static final long SOLVER_TIME = 10;
public static final long DOMAIN_LIMIT = 5000;

public static void main(String[] args) {
  CpModel model = new CpModel();

  for (int i = 0; i < numIter; i++) {
    generateConstraints(model, i%200);
  }
  CpSolver solver = new CpSolver();
  solver.getParameters().setMaxTimeInSeconds(SOLVER_TIME);		 
  solver.getParameters().setMaxDeterministicTime(SOLVER_TIME);		 
  solver.getParameters().setNumSearchWorkers(4);
  solver.solve(model);

  System.out.println(solver.responseStats());		
  model.getBuilder().clear();
  solver.getParameters().clear();
}
	
public static void generateConstraints(CpModel model, int i) {
  IntVar var1 = model.newIntVar(-DOMAIN_LIMIT / 2, DOMAIN_LIMIT / 2, "Var");
  IntVar var2 = model.newIntVar(-DOMAIN_LIMIT / 2, DOMAIN_LIMIT / 2, "Var");
  IntVar varSum = model.newIntVar(-DOMAIN_LIMIT, DOMAIN_LIMIT, "Var");
  IntVar booleanVar = model.newBoolVar("Var");
	
  model.addLinearSumEqual(new IntVar[] {var1, var2}, varSum);
  model.addGreaterOrEqual(varSum, i).onlyEnforceIf(booleanVar);
  model.addLessOrEqual(varSum, i).onlyEnforceIf(booleanVar.not());
}
```
The solver solves the problem in my computer (CPU Intel i5-4460 3.20GHz, RAM 8GB ) in 17 seconds using 4 workers. If only one worker is used, the problem is solved in 23 seconds. Either using 1 or 4 workers the setMaxTimeInSeconds and the setMaxDeterministicTime does not work, since the solver ends after the imposed limit (10 seconds).

**Comment by Mizux** [comment]
Are you sure `solver.getParameters()` return a reference of the internal object and not a copy ?

**Comment by pfolch** [comment]
I made a mistake when I paste the code. I deleted the line that makes the solver solve the model `solver.solve(model)`. 

Anyway, based on the given example in https://developers.google.com/optimization/cp/cp_tasks, the methodology I used to set the parameters is fine (and any other is provided I think), so I guess that `solver.getParameters()` returns a reference of the internal object and not a copy.

**Comment by Mizux** [comment]
can you try to add:
```java
solver.getParameters().setLogSearchProgress(true);
```

FYI if using the sample 
https://github.com/google/or-tools/blob/stable/ortools/sat/samples/SearchForAllSolutionsSampleSat.java
and adding line 72
```java
    solver.getParameters().setMaxTimeInSeconds(5.0);
    solver.getParameters().setLogSearchProgress(true);
```
I got
```sh
make run SOURCE=ortools/sat/samples/SearchForAllSolutionsSampleSat.java
...
#11468   5.00s   num_bool:11472
Solution #11467: time = 5.00 s
  x = 11468
  y = 0
  z = 0
#11469   5.00s   num_bool:11473
Solution #11468: time = 5.00 s
  x = 11469
  y = 0
  z = 0
#11470   5.00s   num_bool:11474
Solution #11469: time = 5.00 s
  x = 11470
  y = 0
  z = 0
11470 solutions found.
```

**Comment by Mizux** [comment]
@pfolch limit must be double...
```java
 public static final double SOLVER_TIME = 10.0;
```

note: on my machine your example take:
```
walltime: 2.95962
usertime: 2.95962
deterministic_time: 0.0279982
```
but if i set to
```java
    solver.getParameters().setLogSearchProgress(true);
    solver.getParameters().setMaxTimeInSeconds(0.9);
```
it seems to stop earlier...

**Comment by pfolch** [comment]
Hi Mizux,

I tried to use the `solver.getParameters().setLogSearchProgress(true)` and the `solver.getParameters().setMaxTimeInSeconds(SOLVER_TIME)` (decalring the SOLVER_TIME constant as a double) and the solver seems to behave as before. Experimenting a little bit, I found that the solver takes into account the time limit if another constraint is added to the model.

```
	public static final long numIter = 20000;
	public static final double SOLVER_TIME = 10.0;
	public static final long DOMAIN_LIMIT = 5000;

	public static void main(String[] args) {
		
		CpModel model = new CpModel();

		for (int i = 0; i < numIter; i++)
			generateConstraints(model, i%200);
		
		CpSolver solver = new CpSolver();
		
//		solver.getParameters().setNumSearchWorkers(4);
		
		solver.getParameters().setLogSearchProgress(true);
		
		solver.getParameters().setMaxTimeInSeconds(SOLVER_TIME);

		solver.solve(model);
		
		System.out.println(solver.responseStats());
		
		model.getBuilder().clear();
		solver.getParameters().clear();

	}
	
	public static void generateConstraints(CpModel model, int i) {
		
		IntVar var1 = model.newIntVar(-DOMAIN_LIMIT / 2, DOMAIN_LIMIT / 2, "Var");
		IntVar var2 = model.newIntVar(-DOMAIN_LIMIT / 2, DOMAIN_LIMIT / 2, "Var");
		IntVar varSum = model.newIntVar(-DOMAIN_LIMIT, DOMAIN_LIMIT, "Var");
		
		IntVar booleanVar = model.newBoolVar("Var");
		
//		model.addMaxEquality(var1, new IntVar[] {var1, var2});
		
//		model.addAllDifferent(new IntVar[] {var1, var2, varSum});
		
		model.addLinearSumEqual(new IntVar[] {var1, var2}, varSum);
		
		model.addGreaterOrEqual(varSum, i).onlyEnforceIf(booleanVar);
		
		model.addLessOrEqual(varSum, i).onlyEnforceIf(booleanVar.not());
		
	}

```
If one of the commented constraints is added to the model, then the solver after 10 seconds stops, and the solver status is UNKOWN. 

Note: If you want to reproduce in your computer more or less the solving time I have running the above example, set the `numIter ` to 50000 and the `DOMAIN_LIMIT` to 10000. It seems that my computer is quite slow...

Note2: The example above (with the commented lines) is the same as the example I shared in the begining, only changing the time (from long to double) and adding the `setLogSearchProgress(true)`.

Thanks for your time and patience.

**Comment by pfolch** [comment]
Sorry, I forgot something, the experiments I did and the results obtained are the following:
1-)  Executing the code as it is pasted in the previous comment (without uncommenting any lines). Result: FEASIBLE, walltime: 24.0577.
2-) Executing the code as it is pasted in the previous comment, but uncommenting the addMaxEquality constraint. Result: UNKNOWN, walltime: 10.0155.
3-) Executing the code as it is pasted in the previous comment, but uncommenting the addAllDifferent constraint. Result: UNKNOWN,walltime: 10.0158.

All these experiments were carried out without uncommenting the setNumSearchWorkers line, which means that they were ran without parallelization. 

The same experiments but using parallelization (with 4 workers), the results are:
1-) CpSolverResponse: status: FEASIBLE, walltime: 17.3527
2-) CpSolverResponse: status: UNKNOWN, walltime: 15.2219
3-) CpSolverResponse: status: UNKNOWN, walltime: 15.1633

Seems that using parallelization the solver behaves a little bit different. The status for each experiment is the same. But for the experiments 2 and 3, the internal time counter does not seem to be updated correctly, and the solver stops at 15 seconds instead to at 10 seconds.

**Comment by lperron** [comment]
I will have a look.

Thanks for the report.

**Comment by lperron** [comment]
So in mono thread, on master, I get (1 worker, 1 second)

    CpSolverResponse:
    status: UNKNOWN
    objective: NA
    best_bound: NA
    booleans: 60000
    conflicts: 0
    branches: 0
    propagations: 0
    integer_propagations: 0
    walltime: 1.03935
    usertime: 1.03935
    deterministic_time: 0
    primal_integral: 0

which is OK

With 4 workers, it returns after 2.4s. I will trace it more thoroughly. 

Thanks again.

**Comment by LMalinge** [comment]
Hi ! I don't know if I'm in the good place but I had a similar problem with my model. I use CpSolver from cp_model from ortools.sat.python.

I tried different time limit : 10 seconds and 60 seconds. With the 10 seconds one, I always get a solution (not the optimal but at least a feasible one) or it stops. I tried to put 60 seconds in the hope of having a better result. Sometimes it works (I mean it stops with or without a solution) but sometimes it never stops ! 

So I wonder if it's a problem with my model or if it's a bug.

**Comment by lperron** [comment]
The time limit has zero effect on the search (apart from stopping it).
I guess the solver just has a hard time finding a first solution. 

Without a model, we cannot investigate.

**Comment by LMalinge** [comment]
Yes I understand.. My problem is that it doesn't stop...

How can I join my binary file ? My model is quite big ...

**Comment by lperron** [comment]
can you copy paste the log ?
Laurent Perron | Operations Research | ***@***.*** | (33) 1 42 68 53
00

**Comment by LMalinge** [comment]
```
Starting CP-SAT solver.
Parameters: max_time_in_seconds: 60 log_search_progress: true

Initial optimization model '':
#Variables: 84611 (1483 in objective)
  - 516 different domains in [0,38846] with a largest complexity of 67.
  - 595 constants in {-1504,-1472,-1445,-1433,-143 ... 9,81,85,87,88,90,91,92,94,95,98,100,108,110,112}
#kBoolOr: 79002 (#enforced: 52668) (#literals: 131670)
#kIntMax: 1
#kInterval: 1482
#kLinear1: 105818 (#enforced: 105336)
#kLinear2: 28487
#kNoOverlap: 149

Starting presolve at 0.05s
[Probing] deterministic_time: 0.348668 (limit: 1) wall_time: 0.470819 (79084/79084)
[Probing]  - new integer holes: 547
[Probing]  - new binary clause: 3476
[SAT presolve] num removable Booleans: 73567 / 84611
[SAT presolve] num trivial clauses: 6801
[SAT presolve] [0s] clauses:4968 literals:9936 vars:3805 one_side_vars:3805 simple_definition:0 singleton_clauses:0
[SAT presolve] [0.00154178s] clauses:1916 literals:3832 vars:3805 one_side_vars:3805 simple_definition:0 singleton_clauses:0
[SAT presolve] [0.00710376s] clauses:1916 literals:3832 vars:3805 one_side_vars:3805 simple_definition:0 singleton_clauses:0
[Probing] deterministic_time: 0.0296906 (limit: 1) wall_time: 0.157349 (79084/79084)
[Probing]  - new binary clause: 539
[SAT presolve] num removable Booleans: 73567 / 84611
[SAT presolve] num trivial clauses: 0
[SAT presolve] [0s] clauses:1916 literals:3832 vars:3805 one_side_vars:3805 simple_definition:0 singleton_clauses:0
[SAT presolve] [0.000762773s] clauses:1916 literals:3832 vars:3805 one_side_vars:3805 simple_definition:0 singleton_clauses:0
[SAT presolve] [0.0039562s] clauses:1916 literals:3832 vars:3805 one_side_vars:3805 simple_definition:0 singleton_clauses:0
712 affine relations still in the model.

Presolve summary:
  - 36011 affine relations were detected.
  - 28445 variable equivalence relations were detected.
  - rule 'affine: fixed' was applied 9 times.
  - rule 'affine: new relation' was applied 17979 times.
  - rule 'bool_and: always false' was applied 13059 times.
  - rule 'bool_or: always true' was applied 12130 times.
  - rule 'bool_or: implications' was applied 66872 times.
  - rule 'bool_or: only one literal' was applied 35367 times.
  - rule 'deductions: 87440 stored' was applied 1 time.
  - rule 'dual: fix variable' was applied 11 times.
  - rule 'dual: reduced domain' was applied 845 times.
  - rule 'enforcement literal not used' was applied 5730 times.
  - rule 'false enforcement literal' was applied 43136 times.
  - rule 'int_max: reduced domains' was applied 1 time.
  - rule 'int_max: removed dup' was applied 1 time.
  - rule 'int_max: removed variables' was applied 1 time.
  - rule 'interval: reduced domains' was applied 4206 times.
  - rule 'interval: unused, converted to linear' was applied 129 times.
  - rule 'linear: always true' was applied 14483 times.
  - rule 'linear: divide by GCD' was applied 14032 times.
  - rule 'linear: dual fixing' was applied 11 times.
  - rule 'linear: empty' was applied 11523 times.
  - rule 'linear: fixed or dup variables' was applied 26457 times.
  - rule 'linear: infeasible' was applied 22308 times.
  - rule 'linear: reduced variable domains' was applied 2944 times.
  - rule 'linear: remapped using affine relations' was applied 14947 times.
  - rule 'linear: simplified rhs' was applied 202600 times.
  - rule 'linear: size one' was applied 29934 times.
  - rule 'linear: tightened into equality' was applied 634 times.
  - rule 'no_overlap: no intervals' was applied 41 times.
  - rule 'no_overlap: removed redundant intervals' was applied 23 times.
  - rule 'objective: variable not used elsewhere' was applied 13 times.
  - rule 'presolve: iteration' was applied 2 times.
  - rule 'removed duplicate constraints' was applied 28708 times.
  - rule 'true enforcement literal' was applied 14647 times.
  - rule 'variables: detect fully reified value encoding' was applied 302 times.
  - rule 'variables: detect half reified value encoding' was applied 602 times.
  - rule 'variables: merge equivalent var value encoding literals' was applied 173 times.

Presolved optimization model '':
#Variables: 10917 (1470 in objective)
  - 1870 different domains in [0,38846] with a largest complexity of 59.
  - 50 constants in {1,2,3,4,5,6,7,8,9,10,11,12,13 ... ,42,44,47,49,50,54,57,58,60,64,68,70,90,110,112}
#kBoolAnd: 1894 (#enforced: 1894) (#literals: 1916)
#kIntMax: 1
#kInterval: 1353
#kLinear1: 18368 (#enforced: 18368)
#kLinear2: 2218
#kNoOverlap: 108

Preloading model.
#Bound   3.79s best:inf   next:[2071329,567942156] initial_domain

Starting to load the model at 3.79s
[Probing] deterministic_time: 0.031724 (limit: 1) wall_time: 0.187781 (6496/6496)
[Probing]  - new integer holes: 1038
[Probing]  - new binary clause: 2772
Initial num_bool: 6496

Starting sequential search at 4.10s
#1       4.61s best:2266189 next:[2071329,2266188]  fixed_bools:1/8686
```

**Comment by lperron** [comment]
Does not seem that big. If you compress the proto file, you should be able
to attach it to the issue, or send it to me by email.
Laurent Perron | Operations Research | ***@***.*** | (33) 1 42 68 53
00

**Comment by LMalinge** [comment]
Oh, ok, thank you !

[model_txt_lmalinge.zip](https://github.com/google/or-tools/files/6677668/model_txt_lmalinge.zip)

**Comment by LMalinge** [comment]
I worked on my model and I think that it worked better now. Thanks for your time :)

**Comment by jmdvinodjmd** [comment]
Hi, I am not sure if I am at right place but is it possible to set time limits and display progress for Cp_model in Python? thank you!

**Comment by lperron** [comment]
what do you mean ?
Laurent Perron | Operations Research | ***@***.*** | (33) 1 42 68 53
00

**Comment by jmdvinodjmd** [comment]
I apologise for the confusion. I am using Cp_model in python to solve MILP. I can't find any way to set time limit on solver so that the solver stops after that time and provides approximate solution. Secondly, when running the code, I can't see intermediate results like Gurobi etc., rather it shows only final results. So, as my program is taking several hours to solve so I am not sure if it is making progress or not.

From the above disscussion it appears, one can set time limits in JAVA. So, can we do also  in Python?

**Comment by Mizux** [comment]
https://developers.google.com/optimization/cp/cp_tasks#time-limit ? see python code
    # Sets a time limit of 10 seconds.
    solver.parameters.max_time_in_seconds = 10.0

**Comment by jmdvinodjmd** [comment]
Thank you! I apologise, I don't know how I missed that, it was right there.

**Comment by cyclotron3k** [comment]
Was this resolved? I have the same issue.
Tested with `com.google.ortools-java:9.12.4544` and `9.14.6206`.

EDIT: I think I was mistaken. It seems the issue is fixed in `9.14.6206`

**Comment by lperron** [comment]
yes, with some parameter tweaks
Laurent Perron | Operations Research | ***@***.*** | (33) 1 42 68 53
00



Le jeu. 26 juin 2025 à 03:50, Aidan Samuel ***@***.***> a
écrit :
